From 0cb9b3736c470f3588e9afff5a4ba07cb0b00cdf Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Tue, 9 Aug 2016 11:43:37 +0100
Subject: [PATCH 02/44] Fix program_options for Boost >= 1.59

You need to use '--var=value' to avoid running into:

'too many positional options have been specified on the command line'

Refs:
http://lists.boost.org/boost-users/2015/09/85130.php
https://github.com/boostorg/program_options/commit/88dea3
---
 src/cpp/core/system/PosixSystem.cpp        |  5 +++--
 src/cpp/desktop-mac/SessionLauncher.mm     |  6 ++----
 src/cpp/desktop/DesktopSessionLauncher.cpp | 16 ++++++----------
 3 files changed, 11 insertions(+), 16 deletions(-)

diff --git a/src/cpp/core/system/PosixSystem.cpp b/src/cpp/core/system/PosixSystem.cpp
index 868a122783..45e5a8c246 100644
--- a/src/cpp/core/system/PosixSystem.cpp
+++ b/src/cpp/core/system/PosixSystem.cpp
@@ -1810,9 +1810,10 @@ Error launchChildProcess(std::string path,
            it != config.args.end();
            ++it)
       {
-         argVector.push_back(it->first);
          if (!it->second.empty())
-            argVector.push_back(it->second);
+            argVector.push_back(it->first + "=" + it->second);
+         else
+            argVector.push_back(it->first);
       }
 
       // allocate ProcessArgs on heap so memory stays around after we exec
diff --git a/src/cpp/desktop-mac/SessionLauncher.mm b/src/cpp/desktop-mac/SessionLauncher.mm
index c31c3e1f32..cd20674677 100644
--- a/src/cpp/desktop-mac/SessionLauncher.mm
+++ b/src/cpp/desktop-mac/SessionLauncher.mm
@@ -121,8 +121,7 @@ Error SessionLauncher::launchFirstSession(const std::string& filename)
    buildLaunchContext(&host, &port, &argList, &appUrl);
 
    // show help home for first session
-   argList.push_back("--show-help-home");
-   argList.push_back("1");
+   argList.push_back("--show-help-home=1");
    
    RUN_DIAGNOSTICS_LOG("\nAttempting to launch R session...");
    logEnvVar("RSTUDIO_WHICH_R");
@@ -349,8 +348,7 @@ void SessionLauncher::buildLaunchContext(std::string* pHost,
    
    if (desktop::options().runDiagnostics())
    {
-      pArgList->push_back("--verify-installation");
-      pArgList->push_back("1");
+      pArgList->push_back("--verify-installation=1");
    }
 }
    
diff --git a/src/cpp/desktop/DesktopSessionLauncher.cpp b/src/cpp/desktop/DesktopSessionLauncher.cpp
index 3609d637c1..6d1c49e217 100644
--- a/src/cpp/desktop/DesktopSessionLauncher.cpp
+++ b/src/cpp/desktop/DesktopSessionLauncher.cpp
@@ -87,8 +87,7 @@ Error SessionLauncher::launchFirstSession(const QString& filename,
    buildLaunchContext(&host, &port, &argList, &url);
 
    // show help home on first run
-   argList.push_back(QString::fromUtf8("--show-help-home"));
-   argList.push_back(QString::fromUtf8("1"));
+   argList.push_back(QString::fromUtf8("--show-help-home=1"));
 
    RUN_DIAGNOSTICS_LOG("\nAttempting to launch R session...");
    logEnvVar("RSTUDIO_WHICH_R");
@@ -378,7 +377,7 @@ void SessionLauncher::buildLaunchContext(QString* pHost,
 
    if (!confPath_.empty())
    {
-      *pArgList << QString::fromUtf8("--config-file") <<
+      *pArgList << QString::fromUtf8("--config-file=") +
                    QString::fromUtf8(confPath_.absolutePath().c_str());
    }
    else
@@ -386,14 +385,12 @@ void SessionLauncher::buildLaunchContext(QString* pHost,
       // explicitly pass "none" so that rsession doesn't read an
       // /etc/rstudio/rsession.conf file which may be sitting around
       // from a previous configuratin or install
-      *pArgList << QString::fromUtf8("--config-file") <<
-                   QString::fromUtf8("none");
+      *pArgList << QString::fromUtf8("--config-file=none");
    }
 
-   *pArgList << QString::fromUtf8("--program-mode") <<
-                QString::fromUtf8("desktop");
+   *pArgList << QString::fromUtf8("--program-mode=desktop");
 
-   *pArgList << QString::fromUtf8("--www-port") << *pPort;
+   *pArgList << QString::fromUtf8("--www-port=") + *pPort;
 
    // create launch token if we haven't already
    if (s_launcherToken.empty())
@@ -402,8 +399,7 @@ void SessionLauncher::buildLaunchContext(QString* pHost,
                 QString::fromUtf8(s_launcherToken.c_str());
    
    if (options().runDiagnostics())
-      *pArgList << QString::fromUtf8("--verify-installation") <<
-                   QString::fromUtf8("1");
+      *pArgList << QString::fromUtf8("--verify-installation=1");
 }
 
 
-- 
2.14.1

